src/chain/candidate.rs

```diff
@@ -0,0 +1,141 @@
+        old_public_id: PublicId,
+    },
+    /// We consensused the candidate online. We are waiting for the SectionInfo to consensus
+    /// and this new node to start handling events before allowing a new candidate.
+    ApprovedWaitingSectionInfo { new_pub_id: PublicId },
+}
+
+impl Candidate {
+    /// Return true if already has a candidate
+    pub fn has_resource_proof_candidate(&self) -> bool {
+        *self != Candidate::None
+    }
+
+    /// Forget about the current candidate.
+    pub fn reset_candidate(&mut self) {
+        *self = Candidate::None;
+    }
+
+    /// Forget about the current candidate if it is a member of the given section.
+    pub fn reset_candidate_if_member_of(&mut self, members: &BTreeSet<PublicId>) {
+        if let Candidate::ApprovedWaitingSectionInfo { ref new_pub_id } = self {
+            if members.contains(new_pub_id) {
+                *self = Candidate::None;
+            }
+        }
+    }
+
+    /// Our section decided that the candidate should be selected next.
+    /// Pre-condition: !has_resource_proof_candidate.
+    pub fn accept_as_candidate(
+        &mut self,
+        old_public_id: PublicId,
+        target_interval: XorTargetInterval,
+    ) {
+        if self.has_resource_proof_candidate() {
+            log_or_panic!(
+                LogLevel::Error,
+                "accept_as_candidate when processing one already"
+            );
+        }
+
+        *self = Candidate::AcceptedForResourceProof {
+            old_public_id: old_public_id,
+            target_interval: target_interval,
+        };
+    }
+
+    /// Return the target interval if we are waiting for candidate info for that PublicId.
+    pub fn waiting_for_candidate_interval(
+        &self,
+        old_public_id: &PublicId,
+    ) -> Option<&XorTargetInterval> {
+        if self.waiting_candidate_old_public_id() == Some(old_public_id) {
+            self.target_interval()
+        } else {
+            None
+        }
+    }
+
+    /// Handle consensus on `Online`. Marks the candidate as `ApprovedWaitingSectionInfo`.
+    /// If the candidate was already purged or is unexpected, return false.
+    pub fn handle_candidate_online_event(&mut self, online_payload: &OnlinePayload) -> bool {
+        if self.waiting_candidate_old_public_id() == Some(&online_payload.old_public_id) {
+            // Known candidate was accepted.
+            // Ignore any information that could be different stored in candidate.
+            // Do not accept candidate until we complete our current one.
+            *self = Candidate::ApprovedWaitingSectionInfo {
+                new_pub_id: online_payload.new_public_id,
+            };
+            true
+        } else {
+            // Unkwown candidate, Candidate was purged before: refuse it.
+            false
+        }
+    }
+
+    /// The public id of the candidate we are waiting to approve.
+    pub fn waiting_candidate_old_public_id(&self) -> Option<&PublicId> {
```

Comment:
This method might be confusingly named. "waiting candidate" to me suggest it refers to the `ApprovedWaitingSectionInfo` variant, but that variants yields `None`. I think it would be renamed to `accepted_candidate_old_public_id` (or just `accepted_old_public_id`).

Reply:
sounds good.